File: TP2/algo2.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int rechElets_TabTries(int tab[], int n, int x) {
    for (int i = 0; i < n; i++) {
        if (tab[i] == x)
            return 1;
        else if (tab[i] > x)
            return 0;
    }
    return 0;
}

double get_time_diff(struct timespec start, struct timespec end) {
    return (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
}

int main() {
    FILE *f_tailles = fopen("tailles.txt", "r");
    FILE *f_res = fopen("resultats_trie.csv", "w");
    if (!f_tailles || !f_res) {
        printf("Erreur lors de l'ouverture d'un fichier.\n");
        return 1;
    }

    fprintf(f_res, "n,temps_meilleur,temps_pire\n");

    int n;
    srand(time(NULL));

    while (fscanf(f_tailles, "%d", &n) == 1) {
        int *tab = (int*) malloc(n * sizeof(int));
        if (!tab) continue;

        for (int i = 0; i < n; i++)
            tab[i] = i * 2;

        struct timespec start, end;
        double temps_meilleur = 0, temps_pire = 0;
        int repetitions = 1000;  // répéter pour fiabiliser les mesures

        // Meilleur cas
        int x_meilleur = tab[0];
        for (int i = 0; i < repetitions; i++) {
            clock_gettime(CLOCK_MONOTONIC, &start);
            rechElets_TabTries(tab, n, x_meilleur);
            clock_gettime(CLOCK_MONOTONIC, &end);
            temps_meilleur += get_time_diff(start, end);
        }
        temps_meilleur /= repetitions;

        // Pire cas
        int x_pire = -1;
        for (int i = 0; i < repetitions; i++) {
            clock_gettime(CLOCK_MONOTONIC, &start);
            rechElets_TabTries(tab, n, x_pire);
            clock_gettime(CLOCK_MONOTONIC, &end);
            temps_pire += get_time_diff(start, end);
        }
        temps_pire /= repetitions;

        fprintf(f_res, "%d,%.10f,%.10f\n", n, temps_meilleur, temps_pire);
        free(tab);
    }

    fclose(f_tailles);
    fclose(f_res);
    printf("Mesures terminées avec précision nanoseconde !\n");
    return 0;
}

File: TP2/algo2.py
import time
import csv

#  Fonction de recherche séquentielle optimisée pour tableau trié
def rechElets_TabTries(tab, x):
    for val in tab:
        if val == x:
            return True
        elif val > x:
            return False  # inutile de continuer
    return False


#  Lecture des tailles + enregistrement des résultats
with open("tailles.txt", "r") as f_tailles, open("resultats_trie.csv", "w", newline="") as f_res:
    writer = csv.writer(f_res)
    writer.writerow(["n", "temps_meilleur", "temps_pire"])

    for line in f_tailles:
        n = int(line.strip())

        #  Génération d’un tableau trié croissant
        tab = [i * 2 for i in range(n)]

        #  Meilleur cas (élément au début)
        x_meilleur = tab[0]
        start = time.perf_counter()
        rechElets_TabTries(tab, x_meilleur)
        end = time.perf_counter()
        temps_meilleur = end - start

        #  Pire cas (élément absent)
        x_pire = tab[-1] + 1
        start = time.perf_counter()
        rechElets_TabTries(tab, x_pire)
        end = time.perf_counter()
        temps_pire = end - start

        #  Sauvegarde dans le CSV
        writer.writerow([n, temps_meilleur, temps_pire])

print("✅ Mesures terminées. Résultats enregistrés dans 'resultats_trie.csv'")

